spacy_llm/tasks/srl_task.py

```diff
@@ -0,0 +1,407 @@
+        self._check_extensions()
+
+        examples = get_examples()
+
+        if not labels:
+            labels = list(self._label_dict.values())
+
+        if not labels:
+            label_set = set()
+
+            for eg in examples:
+                rels: List[ArgRELItem] = eg.reference._.relations
+                for rel in rels:
+                    label_set.add(rel.label)
+            labels = list(label_set)
+
+        self._label_dict = {self._normalizer(label): label for label in labels}
+
+    @property
+    def labels(self) -> Tuple[str, ...]:
+        return tuple(self._label_dict.values())
+
+    @property
+    def prompt_template(self) -> str:
+        return self._template
+
+    def generate_prompts(self, docs: Iterable[Doc]) -> Iterable[str]:
+        environment = jinja2.Environment()
+        _template = environment.from_string(self._template)
+        for doc in docs:
+            predicates = None
+            if len(doc._.predicates):
+                predicates = ", ".join([p["text"] for p in doc._.predicates])
+
+            prompt = _template.render(
+                text=doc.text,
+                labels=list(self._label_dict.values()),
+                label_definitions=self._label_definitions,
+                predicates=predicates,
+            )
+
+            yield prompt
+
+    def _format_response(self, arg_lines):
+        """Parse raw string response into a structured format"""
+        output = []
+        # this ensures unique arguments in the sentence for a predicate
+        found_labels = set()
+        for line in arg_lines:
+            try:
+                if line.strip() and ":" in line:
+                    label, phrase = line.strip().split(":", 1)
+
+                    # label is of the form "ARG-n (def)"
+                    label = label.split("(")[0].strip()
+
+                    # strip any surrounding quotes
+                    phrase = phrase.strip("'\" -")
+
+                    norm_label = self._normalizer(label)
+                    if (
+                        norm_label in self._label_dict
+                        and norm_label not in found_labels
+                    ):
+                        if phrase.strip():
+                            _phrase = phrase.strip()
+                            found_labels.add(norm_label)
+                            output.append((self._label_dict[norm_label], _phrase))
+            except ValidationError:
+                msg.warn(
+                    "Validation issue",
+                    line,
+                    show=self._verbose,
+                )
+        return output
+
+    def parse_responses(
+        self, docs: Iterable[Doc], responses: Iterable[str]
+    ) -> Iterable[Doc]:
+        for doc, prompt_response in zip(docs, responses):
+            predicates = []
+            relations = []
+            lines = prompt_response.split("\n")
+
+            # match lines that start with {Predicate:, Predicate 1:, Predicate1:}
```

Comment:
This behavior should be more clearly documented.

Reply:
I guess it might become more clear if I provide an example of an expected response here or in the README somewhere. Is that what you were going for?

Reply:
I think it'd best if the documentation for this task specifies which lines are parsed properly. This would be helpful if users have to debug LLM output.

Reply:
I am still not clear on how and where I should add the docs for this. Let's take an example and think through:

```sh
LLM response for doc: "A sentence with multiple predicates (p1, p2)"

Step 1: Extract the Predicates for the Text
Predicates: p1, p2

Step 2: For each Predicate, extract the Semantic Roles in 'Text'
Text: A sentence with multiple predicates (p1, p2)
Predicate: p1
ARG-0: a0_1
ARG-1: a1_1
ARG-M-TMP: a_t_1
ARG-M-LOC: a_l_1

Predicate: p2
ARG-0: a0_2
ARG-1: a1_2
ARG-M-TMP: a_t_2
```

So the steps in the parsing are to first find the text boundaries for the information of each predicate. This is by identifying the lines `Predicate: p1` and `Predicate: p2`, which gives us the text for each predicate:

```sh
Predicate: p1
ARG-0: a0_1
ARG-1: a1_1
ARG-M-TMP: a_t_1
ARG-M-LOC: a_l_1
```
and,
```sh
Predicate: p2
ARG-0: a0_2
ARG-1: a1_2
ARG-M-TMP: a_t_2
```

Once we separate these out, then it is a matter of parsing line by line to extract the predicate and its args for each predicate block

Reply:
You can copy this more or less 1:1 into the comments where you deem it appropriate (we can still move it around at some later point - it's just important that we have a clear understanding/documentation of how the parsing works in general).

Reply:
done. https://github.com/explosion/spacy-llm/pull/226/commits/be5065570bd235e5b297f85228770cd100c49ff8